### ISSUE TITLE
Docker Multi-GPU DDP training hang on destroy_process_group()

### ISSUE THREAD

**Issue by Howe-Young** [ultralytics/yolov5#7307]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Multi-GPU

### Bug

When I training yolov5 on docker(4 x Tesla V100), the program will be hang on dist.destroy_process_group(). But it seems that some GPU memory has been released, like this（note that due to the docker configuration problem, the python process id is not displayed, but the process actually exists）: 
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.51.06    Driver Version: 450.51.06    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  On   | 00000000:04:00.0 Off |                    0 |
| N/A   32C    P0    33W / 250W |   1409MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  On   | 00000000:05:00.0 Off |                    0 |
| N/A   29C    P0    24W / 250W |      4MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-PCIE...  On   | 00000000:08:00.0 Off |                    0 |
| N/A   30C    P0    27W / 250W |      4MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-PCIE...  On   | 00000000:09:00.0 Off |                    0 |
| N/A   29C    P0    22W / 250W |      4MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

### Environment

PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-862.mt20190308.130.el7.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
GPU 2: Tesla V100-PCIE-16GB
GPU 3: Tesla V100-PCIE-16GB

Nvidia driver version: 450.51.06
cuDNN version: /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.0
[pip3] torchaudio==0.8.0a0+a751e1d
[pip3] torchvision==0.9.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch                   1.8.0           py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] torchaudio                0.8.0                      py38    pytorch
[conda] torchvision               0.9.0                py38_cu102    pytorch

### Minimal Reproducible Example

```
python3 -m torch.distributed.launch --master_port 33714 --nproc_per_node 4 train.py --data reproduction_121.yaml --hyp data/hyps/hyp.finetune.yaml --adam --weights=yolov5n_gd0.25_gw0.25_effi.pt --epochs=10 --batch-size=64 --img-size=512
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Howe-Young thanks for the bug report! Usually I direct DDP users to our Docker Image but it seems you are already using it, though I'm confused about the PyTorch version, the most recent images, i.e. within the last month are all on PyTorch 1.11.0, so I would simply pull the latest image and re-run to check for reproducibility there.

We run DDP trainings every week without issue in Docker.

**Comment by sandeepjana** [comment]
@Howe-Young I was also facing the hang issue with multi-GPU DDP training. So I commented out `destroy_process_group()` for now. In fact, the hang was not _in_ the call to destroy_process_group but seems to happen on the exit of the train script. The example given in Pytorch DDP tutorial [here]( https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#initialize-ddp-with-torch-distributed-run-torchrun) does not call this API! I am not running the script in the docker by the way.

**Comment by glenn-jocher** [comment]
@sandeepjana the cleanup() function in the tutorial runs destroy_process_group():
https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#initialize-ddp-with-torch-distributed-run-torchrun

<img width="960" alt="Screen Shot 2022-08-09 at 8 04 51 PM" src="https://user-images.githubusercontent.com/26833433/183727159-c690b865-7825-4861-a0a1-c4388af90db9.png">

**Comment by sandeepjana** [comment]
The sample code exactly at the link https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#initialize-ddp-with-torch-distributed-run-torchrun does not call cleanup(). This code piece is intended to be used with `torchrun`. On the other hand, the sample code at https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#basic-use-case does have a call to cleanup(). But this code uses `mp.spawn()` instead of `torchrun`. So I think the clean up procedures may be different. (or it is a simple miss in the tutorial).

**Comment by glenn-jocher** [comment]
@sandeepjana hmm interesting. Can you try running a short DDP training without the call to destroy_procerss_group() and see all resources are freed up on train end (by viewing nvidia-smi).

**Comment by sandeepjana** [comment]
I ran train.py 10 times in a loop training with COCO dataset on 2 GPU cores. Trained for 1 epoch in each run and called `nvidia-smi` at the end of each run. The GPU memory is freed up (0MiB) even after commenting out `destroy_process_group()` when `torchrun/torch.distributed.run` is used. I could not check for CPU memory leaks as I have a shared machine.

**Comment by glenn-jocher** [comment]
@sandeepjana great! I've added a TODO to test this potential change.

**Comment by Howe-Young** [comment]
> @sandeepjana 
when I removed destroy_process_group()，it exits normally.

**Comment by glenn-jocher** [comment]
@Howe-Young @sandeepjana good news 😃! Your original issue may now be fixed ✅ in PR #8935. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
